http-server/src/main/java/xtdb/query/AJoinDeserializer.java

```diff
@@ -0,0 +1,62 @@
+package xtdb.query;
+
+import clojure.lang.Keyword;
+import clojure.lang.PersistentHashMap;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.deser.std.StdDeserializer;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import xtdb.IllegalArgumentException;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class AJoinDeserializer extends StdDeserializer<Query.AJoin> {
+
+    public AJoinDeserializer() {
+        super(Query.Join.class);
+    }
+
+    private List<ArgSpec> deserializeArgs(ObjectMapper mapper, ArrayNode node) throws JsonProcessingException {
+        List<ArgSpec> res = new ArrayList<>();
+        for (JsonNode argSpecNode : node) {
+            res.add(mapper.treeToValue(argSpecNode, ArgSpec.class));
+        }
+        return res;
+    }
+
+    private List<OutSpec> deserializeBind(ObjectMapper mapper, ArrayNode node) throws JsonProcessingException {
+        List<OutSpec> res = new ArrayList<>();
+        for (JsonNode argSpecNode : node) {
+            res.add(mapper.treeToValue(argSpecNode, OutSpec.class));
+        }
+        return res;
+    }
+
+    public Query.AJoin deserialize(JsonParser p, DeserializationContext ctxt) throws IOException, JsonProcessingException {
+        ObjectMapper mapper = (ObjectMapper) p.getCodec();
+        ObjectNode node = mapper.readTree(p);
+
+        try {
+            Query.AJoin join = null;
+            if (node.has("join")) {
+                join = Query.join(mapper.treeToValue(node.get("join"), Query.class), deserializeArgs(mapper, (ArrayNode) node.get("args")));
+            } else {
```

Comment:
wouldn't necessarily assume `node.has("left_join")` here, I don't think? could we specify this in an `else if` and then throw an error in the case that it's neither?

Reply:
I mainly do it this way because the calling class assures there is one or the other. See `UnifyClauseDeser`.